Automattic/vip-cli
# FORNO-665: Public facing Media Import

## Description

This PR adds a new command that gives VIP customers the ability to import Media files into their VIP Go application. 

The tool takes a web URL as input which must be publically accessible and contains an archive of the files that need to be imported.

The tool detects the root folder structure of the archive and looks for the first occurence of the `/uploads` directory and imports all the files after it.

Any files that are intentionally skipped or fail to be imported into the site, are logged and exported to the user in the form of a plain text file.

#### Options

- Export file errors to a json file instead of a plain text file: `-E`
- Import intermediate image files: `-i`
- Overwrite existing files: `-o`

#### Constraints

- Imports can be carried out only on Production environments of VIP Go applications
- Only Web Archives are allowed ( no local files )
- Archives can be of a maximum of 30 GB in size
- Imports can be carried out only on WordPress application of VIP Go applications
- Importing an archive that is hosted in Google Drive isn't supported

## Steps to Test

1. Check out PR.
1. Run `npm run build`
1. Ensure that your local GOOP, Parker and [MMS](https://github.com/Automattic/vip-go-media-management-service#local-development) are running and connected
2. Run the following command:
```bash
VIP_PROXY="" API_HOST=http://localhost:4000 node ./dist/bin/vip-import-media.js @70.production http://transfer.sh/BenQY/test.zip
```
5. Ensure that the command runs successfully and that the import is completed with the likeness of the following output:

```
→ VIP_PROXY="" API_HOST=http://localhost:4000 node ./dist/bin/vip-import-media.js @70.production http://transfer.sh/BenQY/test.zip
===================================
+ App: test-ms-subdir-01-vipv2-net (id: 70)
+ Environment: production (id: 70)
+ Archive URL: http://transfer.sh/BenQY/test.zip
+ Overwrite any existing files: x No
+ Import intermediate image files: x No
+ Export any file errors encountered to a JSON file instead of a plain text file: x No
===================================
✔ 
NOTE: If the provided archive's directory structure contains an `/uploads` directory,
only the files present inside that directory will be imported and the rest will be ignored.
If no `/uploads` directory is found, all files will be imported, as is.

Are you sure you want to import the contents of the url?
 (y/N) · true

Importing archive from: http://transfer.sh/BenQY/test.zip
to: testingmapped.com (PRODUCTION)

=============================================================
Processing the files import for your environment...
Imported Files: 1/2 - 50% ✓
=============================================================
Status: COMPLETED ✓ : The Imported files should be visible on your site testingmapped.com
App: test-ms-subdir-01-vipv2-net (PRODUCTION)
=============================================================

⚠️ 1 file error(s) found
All errors have been exported to /Users/saroshaga/Projects/Automattic/vip/media-import-test-ms-subdir-01-vipv2-net-1621859354156.txt
```

6. You can also check the status of an import with the `vip import media status` command

```
→ VIP_PROXY="" API_HOST=http://localhost:4000 node ./dist/bin/vip-import-media-status.js @70.production                                           (media-import|✚1…)

=============================================================
Checking the Media import status for your environment...
Imported Files: 1/2 - 50% ✓
=============================================================
Status: COMPLETED ✓ : The Imported files should be visible on your site testingmapped.com
App: test-ms-subdir-01-vipv2-net (PRODUCTION)
=============================================================

⚠️ 1 file error(s) found
All errors have been exported to /Users/saroshaga/Projects/Automattic/vip/media-import-test-ms-subdir-01-vipv2-net-1621859583457.txt
```
